### PR TITLE
[FLINK-23411][checkpoint] Expose Checkpoint metrics in task level

### PR DESCRIPTION
**Pull request by masteryhx**
## What is the purpose of the change

 - Expose task level checkpoint metircs that just like what shows in Flink UI.

## Brief change log

- Introduce AsyncCheckpointMetricsGroup to tracker CheckpointMetrics related to async phase
- Make subtask checkpoint coordinator register and report CheckpointMetrics
- Expose more sync metrics in InputProcessorUtil

## Verifying this change

This change added tests and can be verified as follows:

- Add AsyncCheckpointMetricsGroupTest to test AsyncCheckpointMetricsGroup
- Add more cases in SourceStreamTaskTestBase

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? no

### CODE REVIEW

**Comment by masteryhx** [comment]
@rkhachatryan @StefanRRichter @reswqa 
Could you help to take a review ? 
Thanks a lot !

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/checkpointing/CheckpointBarrierHandler.java [hunk 1]
```diff
@@ -118,6 +125,18 @@ public long getCheckpointStartDelayNanos() {
...
+    public long getBytesProcessedDuringAlignment() {
+        if (isDuringAlignment()) {
+            return bytesProcessedDuringAlignment;
+        } else {
+            return FutureUtils.getOrDefault(latestBytesProcessedDuringAlignment, 0L);
+        }
+    }
+
+    public long getCheckpointSyncDurationMillis() {
+        return latestCheckpointSyncDurationMillis;
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/checkpointing/CheckpointBarrierHandler.java:137]: nit: rename both field and getter to `last/getLastCheckpointSyncDurationMillis`?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/checkpointing/CheckpointBarrierHandler.java [hunk 2]
```diff
@@ -118,6 +125,18 @@ public long getCheckpointStartDelayNanos() {
...
 
+    public long getBytesProcessedDuringAlignment() {
+        if (isDuringAlignment()) {
+            return bytesProcessedDuringAlignment;
+        } else {
+            return FutureUtils.getOrDefault(latestBytesProcessedDuringAlignment, 0L);
+        }
+    }
+
+    public long getCheckpointSyncDurationMillis() {
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/checkpointing/CheckpointBarrierHandler.java:136]: Both of those metrics are untested as far as I can tell.

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/AsyncCheckpointMetricsGroupTest.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+                        .setBytesPersistedOfThisCheckpoint(34L)
+                        .setTotalBytesPersisted(45L)
+                        .setAsyncDurationMillis(56L)
+                        .setSyncDurationMillis(67L)
+                        .setCheckpointStartDelayNanos(78L)
+                        .build();
+
+        // Update and Verify the status of AsyncCheckpointMetricsGroup.
+        asyncCheckpointMetricsGroup.reportCheckpointMetrics(checkpointMetrics);
+        assertThat(
```
Comment (pnowojski) [inline flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/AsyncCheckpointMetricsGroupTest.java:64]: This test doesn't fully cover newly added code. For example one might remove `asyncCheckpointMetricsGroup.reportCheckpointMetrics()` call from the `AsyncCheckpointRunnable` and this test would be still passing.

To have full test coverage I think you should try to write a metrics reporting test in `SubtaskCheckpointCoordinatorTest`.

**Review by pnowojski — CHANGES_REQUESTED** [review]
Thanks for the contribution. Apart of the inline comments, I think those metrics would have to be hidden behind an opt-in feature flag (as I commented in the ticket)

**Comment by pnowojski** [comment]
I'm closing that as I will be publishing a FLIP counter-proposal that uses tracing to achieve the same goal.

**Comment by masteryhx** [comment]
> I'm closing that as I will be publishing a FLIP counter-proposal that uses tracing to achieve the same goal.

Thanks @pnowojski for helping to push it forward, look forward to the FLIP!

**Pull request closed without merging**